### ISSUE TITLE
shell-local post-processor not running commands

### ISSUE THREAD

**Issue by VMTrooper** [mitchellh/packer#3354]
**Description:**
I am attempting to use the shell-local post-processor to run a glance command after building my image to upload it to OpenStack.

The command never executes. I tried changing to something simple like touch some-file-name.txt. This also does not execute.

**Bug Details:**
- Packer Version - 0.9.0
- Host platform - Ubuntu 15.04, OS X 10.11
- [Debug log output from `PACKER_LOG=1 packer build template.json`](https://gist.github.com/VMTrooper/1028092f186c23e1dd1a)
- [Example Template](https://gist.github.com/VMTrooper/0d19fcaba800f1cc5d63)

NOTE: For simplicity, I used the null builder in my example code, but the post-processor behaves the same whether I use the AWS, OpenStack, or null builders. The command does not seem to be executed.

Any help would be appreciated. Thanks!

**Comment by cbednarski** [comment]
Thanks for opening an issue. The reason this is not running using the other builders is because they don't produce any local files. The logic in the post-processor looks like this:

```
	for _, file := range artifact.Files() {
		for _, path := range scripts {
```

I understand this is a bit confusing but this is working as designed. Post-processors are not a lifecycle hook. Rather, post-processors are meant to operate on the artifact. Since there is no local artifact, there is nothing for shell-local to do.

While we could _probably_ rework shell-local to work with remote artifacts I think this is really messy -- how do we know which artifact to use in openstack, for example? There is a feature request open for a [glance post-processor](https://github.com/mitchellh/packer/issues/2678) which I think is the better way to approach this.

In the mean time, I recommend using make or bash to follow up with additional steps after the packer build succeeds.

So in summary:

- The `shell-local` post-processor doesn't work this way, sorry!
- If you have additional feedback on the glance issue, please add it to #2678

**Comment by VMTrooper** [comment]
Thank you for the information, Chris!

I will try again with the VMware builder to see if I can get a Fusion-produced image uploaded into OpenStack.

That is my desired test case in any case. I just wanted to use the null builder to test the functionality instead of waiting through an OS deployment to test the post-processor.

**Comment by rickard-von-essen** [comment]
@cbednarski I think we need to be clearer on this in the documentation. It wasn't clear even for me from the docs.

**Comment by cbednarski** [comment]
@rickard-von-essen Agreed. shell-local stuff was not part of the core design so there are a lot of edge cases around it, unfortunately.

**Comment by hco** [comment]
I think a documentation on how to work around this problem would be great as well.

I see quite a lot of reasons that you need post-processing for remote artifacts, so it would greatly enhance the UX if users would be provided with a solution.

**Comment by deitch** [comment]
So if it runs for each output file, then it breaks for things like the `virtualbox` builders, because they produce 2 outputs: the `.vmdk` file and the `.ovf` file. It will try to run the script for each, even though there is just one "true" artifact (the `ovf` or the `vmdk`). Is that correct?

Is there any way for it to tell which file it is running on?

**Comment by vtolstov** [comment]
Detect which filename passed to script and simple exit 0 if name of artifact not needed.

**Comment by deitch** [comment]
@vtolstov that makes sense. Just look at `$1`?

**Comment by deitch** [comment]
BTW, packer is fantastic.... but the options and behaviours are poorly documented, if at all. Like this issue with how it executes and how many times it runs. Or that the name of the artifact even is passed as the first argument on the command-line to the script. Etc.

**Comment by vtolstov** [comment]
patches welcome...
P.S. developer not like to write docs ;)

**Comment by deitch** [comment]
@vtolstov if I can find the time, I'd be happy to. I am glad packer has a corporate sponsor. 

> developer not like to write docs ;)

You kidding? I enjoy writing docs! I find it forces me to clarify what, exactly, the end result should be: the API, usage, flows. Docs are like good product management for my work.

**Comment by deitch** [comment]
OK, I am doing a very simple pull request now, just a few things added to the shell-local post-processor docs.

**Comment by cbednarski** [comment]
Hey folks, @deitch updated the docs in https://github.com/mitchellh/packer/commit/e9a9219725d095e7ddca7202100a998e445fd867 so I'm going to close this again. Thanks for the additional feedback!
